python/adjoint/Objective.py

```diff
@@ -0,0 +1,661 @@
+            elif grid_info and self.size==grid_info.size:
+                self.name = 'fullgrid_{}'.format(self.celltype)
+            else:
+                 self.name = '{}_{}'.format(self.celltype, len(self.get_cell_names()))
+        self.add_cell_name(self.name)
+
+        # FIXME At present the 'xyzw' metadata cannot be computed until a mp.simulation / meep::fields
+        #       object has been created, but in fact the metadata only depend on the GridInfo
+        #       (resolution and extents of the computational lattice) and are independent
+        #       of the specific material geometry and source configuration of any particular
+        #       'fields' instance or simulation. In keeping with the spirit of 'DFTCell' it should
+        #       be possible to compute the metadata once and for all right here before any mp.simulation()
+        #       or meep::fields is created, but that will require some code refactoring. For the time being
+        #       we punt on this until later, after a fields object has been created.
+        self.xyzw = self.slice_dims = None
+
+    ######################################################################
+    # 'register' the cell with a MEEP timestepping simulation to request
+    # computation of frequency-domain fields
+    ######################################################################
+    def register(self, sim):
+        self.sim     = sim
+        self.dft_obj =      sim.add_flux(self.fcen,self.df,self.nfreq,self.region) if self.celltype=='flux'   \
+                       else sim.add_dft_fields(self.components, self.freqs[0], self.freqs[-1], self.nfreq, where=self.region)
+
+        # take the opportunity to fill in the metadata if not done yet; #FIXME to be removed as discussed above
+        if self.xyzw is None:
+            self.xyzw  = sim.get_dft_array_metadata(center=self.center, size=self.size)
+            self.slice_dims = np.shape(self.xyzw[3])
+
+
+    ######################################################################
+    # Compute an array of frequency-domain field amplitudes, i.e. a
+    # frequency-domain array slice, for a single field component at a
+    # single frequency in the current simulation. This is like
+    # mp.get_dft_array(), but 'zero-padded:' when the low-level DFT object
+    # does not have data for the requested component (perhaps because it vanishes
+    # identically by symmetry), this routine returns an array of the expected
+    # dimensions with all zero entries, instead of a rank-0 array that prints
+    # out as a single preposterously large or small floating-point number,
+    # which is the not-very-user-friendly behavior of mp.get_dft_array().
+    ######################################################################
+    def get_EH_slice(self, c, nf=0):
+        EH = self.sim.get_dft_array(self.dft_obj, c, nf)
+        return EH if np.ndim(EH)>0 else 0.0j*np.zeros(self.slice_dims)
```

Comment:
It seems like this is a bugfix that should be incorporated into `get_dft_array` itself?